controllers/cluster_controller.go

```diff
@@ -287,3 +296,57 @@ func splitMachineList(list *clusterv1.MachineList) (*clusterv1.MachineList, *clu
 	}
 	return controlplanes, nodes
 }
+
+func (r *ClusterReconciler) reconcileClusterStatus(ctx context.Context, cluster *clusterv1.Cluster) error {
+	if cluster.Status.ControlPlaneInitialized {
+		return nil
+	}
+
+	machines := &clusterv1.MachineList{}
+	err := r.Client.List(ctx, machines, client.InNamespace(cluster.Namespace))
+	if err != nil {
+		return err
+	}
+
+	for _, m := range machines.Items {
+		if util.IsControlPlaneMachine(&m) && m.Status.NodeRef != nil {
+			cluster.Status.ControlPlaneInitialized = true
+			return nil
+		}
+	}
+
+	return nil
+}
+
+// MachineToMachineSets is a handler.ToRequestsFunc to be used to enqeue requests for reconciliation
+// for Cluster to update its status.controlplaneInitialized field
+func (r *ClusterReconciler) MachineToCluster(o handler.MapObject) []ctrl.Request {
+	result := []ctrl.Request{}
+
+	m, ok := o.Object.(*clusterv1.Machine)
+	if !ok {
+		r.Log.Error(errors.New("did not get machine"), "got", fmt.Sprintf("%T", o.Object))
```

Comment:
I'm almost hesitant to suggest this, but I wonder if we should consider a `panic` here - if we don't get the expected type, it's because of programmer error and the implication is that we'll never set ControlPlaneInitialized to true. It's probably easier to diagnose programmer error with a panic vs. digging through logs. Am I crazy? 😄

Reply:
I don't disagree, I'm +1 to use Fatal or panic

Reply:
This is what we are doing in other places, so we'd need to update all the existing use cases as well if we wanted to go down that path. (Not saying we shouldn't, just outside the scope of this PR).